index.js

```diff
@@ -0,0 +1,365 @@
+
+                    this.loadPlugin(scm.plugin, options);
+                } else {
+                    throw new Error(`Display name not specified for ${scm.plugin} scm plugin`);
+                }
+            });
+        }
+
+        if (Object.keys(this.scms).length === 0) {
+            throw new Error('No scm config passed in.');
+        }
+    }
+
+    /**
+     * load scm module
+     * @method loadPlugin
+     * @param  {String}         plugin       load plugin name
+     * @param  {Object}         options      settings for scm module
+     */
+    loadPlugin(plugin, options) {
+        if (plugin === 'router') {
+            console.warn('The plugin of scm-router can not be specified for scms setting');
+
+            return;
+        }
+
+        // eslint-disable-next-line global-require, import/no-dynamic-require
+        const ScmPlugin = require(`screwdriver-scm-${plugin}`);
+        const scmPlugin = new ScmPlugin(options);
+        const scmContexts = scmPlugin.getScmContexts();
+        const scmContext = scmContexts[0]; // plugins return only one scmContext
+
+        if (scmContext == null || scmContext.length === 0) {
+            console.warn(`Illegal scmContext for ${plugin} scm plugin`);
+
+            return;
+        }
+
+        if (typeof this.scms[scmContext] === 'object') {
+            console.warn(`Duplicate scm config for ${scmContext}`);
+
+            return;
+        }
+
+        this.scms[scmContext] = scmPlugin;
+    }
+
+    /**
+     * choose webhook scm module
+     * @method chooseWebhookScm
+     * @param  {Object}     headers          The request headers associated with the webhook payload
+     * @param  {Object}     payload          The webhook payload received from the SCM service
+     * @return {Promise}                     scm object
+     */
+    chooseWebhookScm(headers, payload) {
+        return new Promise((resolve, reject) => {
+            // choose a webhook scm module, or null if there is no suitable one
+            async.detectSeries(this.scms, (scm, cb) => {
+                scm.canHandleWebhook(headers, payload)
+                    .then((result) => {
+                        cb(result === false ? null : scm);
+                    }).catch(() => {
+                        cb(null);
+                    });
+            }, (ret) => {
+                if (ret == null) {
+                    reject('there is no suitable webhook module');
+                }
+
+                resolve(ret);
+            });
+        });
+    }
+
+    /**
+     * choose scm module
+     * @method chooseScm
+     * @param  {Object}     config              Configuration
+     * @param  {String}     config.scmContext   Name of scm context
+     * @return {Promise}                        scm object
+     */
+    chooseScm(config) {
+        return new Promise((resolve, reject) => {
+            if (config.scmContext && typeof config.scmContext === 'string') {
```

Comment:
this pattern keeps coming, we should just define `isObject = val => val && typeof val === 'object'` and then update this if to `if(isObject(config) && config.scmContext === 'string')`